File: lib/autoprovision/project_helper.rb
require 'xcodeproj'
require 'json'
require 'plist'
require 'English'

# ProjectHelper ...
class ProjectHelper
  attr_reader :main_target
  attr_reader :targets
  attr_reader :platform

  def initialize(project_or_workspace_path, scheme_name, configuration_name)
    raise "project not exist at: #{project_or_workspace_path}" unless File.exist?(project_or_workspace_path)

    extname = File.extname(project_or_workspace_path)
    raise "unkown project extension: #{extname}, should be: .xcodeproj or .xcworkspace" unless ['.xcodeproj', '.xcworkspace'].include?(extname)

    @project_path = project_or_workspace_path

    # ensure scheme exist
    scheme, scheme_container_project_path = read_scheme_and_container_project(scheme_name)

    # read scheme application targets
    @main_target, @targets_container_project_path = read_scheme_archivable_target_and_container_project(scheme, scheme_container_project_path)
    @platform = @main_target.platform_name

    @targets = collect_dependent_targets(@main_target)
    raise 'failed to collect scheme targets' if @targets.empty?

    # ensure configuration exist
    action = scheme.archive_action
    raise "archive action not defined for scheme: #{scheme_name}" unless action
    default_configuration_name = action.build_configuration
    raise "archive action's configuration not found for scheme: #{scheme_name}" unless default_configuration_name

    if configuration_name.empty? || configuration_name == default_configuration_name
      @configuration_name = default_configuration_name
    elsif configuration_name != default_configuration_name
      targets.each do |target_obj|
        configuration = target_obj.build_configuration_list.build_configurations.find { |c| configuration_name.to_s == c.name }
        raise "build configuration (#{configuration_name}) not defined for target: #{@main_target.name}" unless configuration
      end

      Log.warn("Using defined build configuration: #{configuration_name} instead of the scheme's default one: #{default_configuration_name}")
      @configuration_name = configuration_name
    end

    @build_settings_by_target = {}
  end

  def uses_xcode_auto_codesigning?
    main_target = @targets[0]

    # target attributes
    target_id = main_target.uuid

    project = Xcodeproj::Project.open(@targets_container_project_path)
    attributes = project.root_object.attributes['TargetAttributes']
    if attributes
      target_attributes = attributes[target_id] || {}
      return true if target_attributes['ProvisioningStyle'] == 'Automatic'
    end

    # target build settings
    main_target.build_configuration_list.build_configurations.each do |build_configuration|
      next unless build_configuration.name == @configuration_name

      build_settings = build_configuration.build_settings
      return true if build_settings['CODE_SIGN_STYLE'] == 'Automatic'
    end

    false
  end

  def project_codesign_identity
    codesign_identity = nil

    @targets.each do |target|
      target_name = target.name

      target_identity = target_codesign_identity(target_name)
      Log.debug("#{target_name} codesign identity: #{target_identity} ")

      if target_identity.to_s.empty?
        Log.warn("no CODE_SIGN_IDENTITY build settings found for target: #{target_name}")
        next
      end

      if codesign_identity.nil?
        codesign_identity = target_identity
        next
      end

      unless codesign_identites_match?(codesign_identity, target_identity)
        Log.warn("target codesign identity: #{target_identity} does not match to the already registered codesign identity: #{codesign_identity}")
        codesign_identity = nil
        break
      end

      codesign_identity = exact_codesign_identity(codesign_identity, target_identity)
    end

    raise 'failed to determine project code sign identity' unless codesign_identity

    codesign_identity
  end

  def project_team_id
    team_id = nil

    project = Xcodeproj::Project.open(@targets_container_project_path)
    attributes = project.root_object.attributes['TargetAttributes'] || {}

    @targets.each do |target|
      target_name = target.name

      current_team_id = target_team_id(target_name)
      Log.debug("#{target_name} target build settings team id: #{current_team_id}")

      unless current_team_id
        Log.warn("no DEVELOPMENT_TEAM build settings found for target: #{target_name}, checking target attributes...")

        target_attributes = attributes[target.uuid] if attributes
        target_attributes_team_id = target_attributes['DevelopmentTeam'] if target_attributes
        Log.debug("#{target_name} target attributes team id: #{target_attributes_team_id}")

        unless target_attributes_team_id
          Log.warn("no DevelopmentTeam target attribute found for target: #{target_name}")
          next
        end

        current_team_id = target_attributes_team_id
      end

      if team_id.nil?
        team_id = current_team_id
        next
      end

      next if team_id == current_team_id

      Log.warn("target team id: #{current_team_id} does not match to the already registered team id: #{team_id}")
      team_id = nil
      break
    end

    team_id
  end

  def target_bundle_id(target_name)
    build_settings = xcodebuild_target_build_settings(target_name)

    bundle_id = build_settings['PRODUCT_BUNDLE_IDENTIFIER']
    return bundle_id if bundle_id

    Log.debug("PRODUCT_BUNDLE_IDENTIFIER env not found in 'xcodebuild -showBuildSettings -project \"#{@targets_container_project_path}\" -target \"#{target_name}\" -configuration \"#{@configuration_name}\"' command's output")
    Log.debug("checking the Info.plist file's CFBundleIdentifier property...")

    info_plist_path = build_settings['INFOPLIST_FILE']
    raise 'failed to to determine bundle id: xcodebuild -showBuildSettings does not contains PRODUCT_BUNDLE_IDENTIFIER nor INFOPLIST_FILE' unless info_plist_path

    info_plist_path = File.expand_path(info_plist_path, File.dirname(@targets_container_project_path))
    info_plist = Plist.parse_xml(info_plist_path)
    bundle_id = info_plist['CFBundleIdentifier']
    raise 'failed to to determine bundle id: xcodebuild -showBuildSettings does not contains PRODUCT_BUNDLE_IDENTIFIER nor Info.plist' if bundle_id.to_s.empty?

    return bundle_id unless bundle_id.to_s.include?('$')

    Log.warn("CFBundleIdentifier defined with variable: #{bundle_id}, trying to resolve it...")
    resolved = resolve_bundle_id(bundle_id, build_settings)
    Log.warn("resolved CFBundleIdentifier: #{resolved}")
    resolved
  end

  def target_entitlements(target_name)
    settings = xcodebuild_target_build_settings(target_name)
    entitlements_path = settings['CODE_SIGN_ENTITLEMENTS']
    return if entitlements_path.to_s.empty?

    project_dir = File.dirname(@targets_container_project_path)
    entitlements_path = File.join(project_dir, entitlements_path)
    Plist.parse_xml(entitlements_path)
  end

  def force_code_sign_properties(target_name, development_team, code_sign_identity, provisioning_profile_uuid)
    target_found = false
    configuration_found = false

    project = Xcodeproj::Project.open(@targets_container_project_path)
    project.targets.each do |target_obj|
      next unless target_obj.name == target_name
      target_found = true

      # force target attributes
      target_id = target_obj.uuid
      attributes = project.root_object.attributes['TargetAttributes']
      if attributes
        target_attributes = attributes[target_id]
        if target_attributes
          target_attributes['ProvisioningStyle'] = 'Manual'
          target_attributes['DevelopmentTeam'] = development_team
          target_attributes['DevelopmentTeamName'] = ''
        end
      end

      # force target build settings
      target_obj.build_configuration_list.build_configurations.each do |build_configuration|
        next unless build_configuration.name == @configuration_name
        configuration_found = true

        build_settings = build_configuration.build_settings
        codesign_settings = {
          'CODE_SIGN_STYLE' => 'Manual',
          'DEVELOPMENT_TEAM' => development_team,

          'CODE_SIGN_IDENTITY' => code_sign_identity,
          'CODE_SIGN_IDENTITY[sdk=iphoneos*]' => code_sign_identity,

          'PROVISIONING_PROFILE_SPECIFIER' => '',
          'PROVISIONING_PROFILE' => provisioning_profile_uuid,
          'PROVISIONING_PROFILE[sdk=iphoneos*]' => provisioning_profile_uuid
        }
        build_settings.merge!(codesign_settings)

        Log.print(JSON.pretty_generate(codesign_settings))
      end
    end

    raise "target (#{target_name}) not found in project: #{@targets_container_project_path}" unless target_found
    raise "configuration (#{@configuration_name}) does not exist in project: #{@targets_container_project_path}" unless configuration_found

    project.save
  end

  private

  def read_scheme_and_container_project(scheme_name)
    project_paths = [@project_path]
    project_paths += contained_projects if workspace?

    project_paths.each do |project_path|
      schema_path = File.join(project_path, 'xcshareddata', 'xcschemes', scheme_name + '.xcscheme')
      next unless File.exist?(schema_path)

      return Xcodeproj::XCScheme.new(schema_path), project_path
    end

    raise "project (#{@project_path}) does not contain scheme: #{scheme_name}"
  end

  def archivable_target_and_container_project(buildable_references, scheme_container_project_dir)
    buildable_references.each do |reference|
      next if reference.target_name.to_s.empty?
      next if reference.target_referenced_container.to_s.empty?

      container = reference.target_referenced_container.sub(/^container:/, '')
      next if container.empty?

      target_project_path = File.expand_path(container, scheme_container_project_dir)
      next unless File.exist?(target_project_path)

      project = Xcodeproj::Project.open(target_project_path)
      target = project.targets.find { |t| t.name == reference.target_name }
      next unless target
      next unless runnable_target?(target)

      return target, target_project_path
    end
  end

  def read_scheme_archivable_target_and_container_project(scheme, scheme_container_project_path)
    build_action = scheme.build_action
    return nil unless build_action

    entries = build_action.entries || []
    return nil if entries.empty?

    entries = entries.select(&:build_for_archiving?) || []
    return nil if entries.empty?

    scheme_container_project_dir = File.dirname(scheme_container_project_path)

    entries.each do |entry|
      buildable_references = entry.buildable_references || []
      next if buildable_references.empty?

      target, target_project_path = archivable_target_and_container_project(buildable_references, scheme_container_project_dir)
      next if target.nil? || target_project_path.nil?

      return target, target_project_path
    end

    raise 'failed to find scheme archivable target'
  end

  def collect_dependent_targets(target, dependent_targets = [])
    dependent_targets << target

    dependencies = target.dependencies || []
    return dependent_targets if dependencies.empty?

    dependencies.each do |dependency|
      dependent_target = dependency.target
      next unless dependent_target
      next unless runnable_target?(dependent_target)

      collect_dependent_targets(dependent_target, dependent_targets)
    end

    dependent_targets
  end

  def target_codesign_identity(target_name)
    settings = xcodebuild_target_build_settings(target_name)
    settings['CODE_SIGN_IDENTITY']
  end

  def target_team_id(target_name)
    settings = xcodebuild_target_build_settings(target_name)
    settings['DEVELOPMENT_TEAM']
  end

  def workspace?
    extname = File.extname(@project_path)
    extname == '.xcworkspace'
  end

  def contained_projects
    return [@project_path] unless workspace?

    workspace = Xcodeproj::Workspace.new_from_xcworkspace(@project_path)
    workspace_dir = File.dirname(@project_path)
    project_paths = []
    workspace.file_references.each do |ref|
      pth = ref.path
      next unless File.extname(pth) == '.xcodeproj'
      next if pth.end_with?('Pods/Pods.xcodeproj')

      project_path = File.expand_path(pth, workspace_dir)
      project_paths << project_path
    end

    project_paths
  end

  def runnable_target?(target)
    return false unless target.is_a?(Xcodeproj::Project::Object::PBXNativeTarget)

    product_reference = target.product_reference
    return false unless product_reference

    product_reference.path.end_with?('.app', '.appex')
  end

  def project_targets_map
    project_targets = {}

    project_paths = contained_projects
    project_paths.each do |project_path|
      targets = []

      project = Xcodeproj::Project.open(project_path)
      project.targets.each do |target|
        next unless runnable_target?(target)

        targets.push(target.name)
      end

      project_targets[project_path] = targets
    end

    project_targets
  end

  def xcodebuild_target_build_settings(target)
    raise 'xcodebuild -showBuildSettings failed: target not specified' if target.to_s.empty?

    settings = @build_settings_by_target[target]
    return settings if settings

    cmd = [
      'xcodebuild',
      '-showBuildSettings',
      '-project',
      "\"#{@targets_container_project_path}\"",
      '-target',
      "\"#{target}\"",
      '-configuration',
      "\"#{@configuration_name}\""
    ].join(' ')

    Log.debug("$ #{cmd}")
    out = `#{cmd}`
    raise "#{cmd} failed, out: #{out}" unless $CHILD_STATUS.success?

    settings = {}
    lines = out.split(/\n/)
    lines.each do |line|
      line = line.strip
      next unless line.include?(' = ')

      split = line.split(' = ')
      next unless split.length == 2

      value = split[1].strip
      next if value.empty?

      key = split[0].strip
      next if key.empty?

      settings[key] = value
    end

    @build_settings_by_target[target] = settings
    settings
  end

  def resolve_bundle_id(bundle_id, build_settings)
    # Bitrise.$(PRODUCT_NAME:rfc1034identifier)
    pattern = /(.*)\$\((.*)\)(.*)/
    matches = bundle_id.match(pattern)
    raise "failed to resolve bundle id (#{bundle_id}): does not conforms to: /(.*)$\(.*\)(.*)/" unless matches

    captures = matches.captures
    prefix = captures[0]
    suffix = captures[2]
    env_key = captures[1]
    split = env_key.split(':')
    raise "failed to resolve bundle id (#{bundle_id}): failed to determine settings key" if split.empty?

    env_key = split[0]
    env_value = build_settings[env_key]
    raise "failed to resolve bundle id (#{bundle_id}): build settings not found with key: (#{env_key})" if env_value.to_s.empty?

    prefix + env_value + suffix
  end

  # 'iPhone Developer' should match to 'iPhone Developer: Bitrise Bot (ABCD)'
  def codesign_identites_match?(identity1, identity2)
    return true if identity1.downcase.include?(identity2.downcase)
    return true if identity2.downcase.include?(identity1.downcase)
    false
  end

  # 'iPhone Developer: Bitrise Bot (ABCD)' is exact compared to 'iPhone Developer'
  def exact_codesign_identity(identity1, identity2)
    return nil unless codesign_identites_match?(identity1, identity2)
    identity1.length > identity2.length ? identity1 : identity2
  end
end

File: autoprovision/projecthelper.go
package autoprovision

import (
	"errors"
	"fmt"
	"path/filepath"
	"regexp"
	"strings"

	"github.com/bitrise-io/go-utils/fileutil"
	"github.com/bitrise-io/go-utils/log"
	"github.com/bitrise-io/go-utils/pathutil"
	"github.com/bitrise-io/go-utils/sliceutil"
	"github.com/bitrise-io/xcode-project/serialized"
	"github.com/bitrise-io/xcode-project/xcodeproj"
	"github.com/bitrise-io/xcode-project/xcscheme"
	"github.com/bitrise-io/xcode-project/xcworkspace"
	"github.com/bitrise-steplib/steps-ios-auto-provision/appstoreconnect"
	"howett.net/plist"
)

// ProjectHelper ...
type ProjectHelper struct {
	MainTarget    xcodeproj.Target
	Targets       []xcodeproj.Target
	Platform      Platform
	XcProj        xcodeproj.XcodeProj
	Configuration string
}

// Platform of the target
// iOS, tvOS, macOS
type Platform string

// BundleIDPlatform ...
func (p Platform) BundleIDPlatform() (*appstoreconnect.BundleIDPlatform, error) {
	var apiPlatform appstoreconnect.BundleIDPlatform
	switch p {
	case IOS, TVOS:
		apiPlatform = appstoreconnect.IOS
	case MacOS:
		apiPlatform = appstoreconnect.MacOS
	default:
		return nil, errors.New("unknown platform: " + string(p))
	}
	return &apiPlatform, nil
}

// Const
const (
	IOS   Platform = "iOS"
	TVOS  Platform = "tvOS"
	MacOS Platform = "macOS"
)

// NewProjectHelper checks the provided project or workspace and generate a ProjectHelper with the provided scheme and configuration
// Previously in the ruby version the initialize method did the same
// It returns a new ProjectHelper pointer and a configuration to use.
func NewProjectHelper(projOrWSPath, schemeName, configurationName string) (*ProjectHelper, string, error) {
	// Maybe we should do this checks during the input parsing
	if exits, err := pathutil.IsPathExists(projOrWSPath); err != nil {
		return nil, "", err
	} else if !exits {
		return nil, "", fmt.Errorf("provided path does not exists: %s", projOrWSPath)
	}

	// Get the project of the provided .xcodeproj or .xcworkspace
	xcproj, _, err := findBuiltProject(projOrWSPath, schemeName, configurationName)
	if err != nil {
		return nil, "", fmt.Errorf("failed to find build project, error: %s", err)
	}

	mainTarget, err := mainTargetOfScheme(xcproj, schemeName)
	if err != nil {
		return nil, "", fmt.Errorf("failed to find the main target of the scheme (%s), error: %s", schemeName, err)
	}

	scheme, ok := xcproj.Scheme(schemeName)
	if !ok {
		return nil, "", fmt.Errorf("no scheme found with name: %s in project: %s", schemeName, projOrWSPath)
	}

	// Check if the archive is availabe for the scheme or not
	if _, archivable := scheme.AppBuildActionEntry(); archivable != true {
		return nil, "", fmt.Errorf("archive action not defined for scheme: %s", scheme.Name)
	}

	// Get the platform (PLATFORM_DISPLAY_NAME) -iphoneos, macosx, appletvos
	platf, err := platform(xcproj, mainTarget, configurationName)
	if err != nil {
		return nil, "", fmt.Errorf("failed to find platform of the project, error: %s", err)
	}

	// Configuration
	conf, err := configuration(configurationName, scheme, xcproj)
	if err != nil {
		return nil, "", err
	}
	return &ProjectHelper{
		MainTarget:    mainTarget,
		Targets:       xcproj.Proj.Targets,
		Platform:      platf,
		XcProj:        xcproj,
		Configuration: conf,
	}, conf,
		nil
}

func (p ProjectHelper) ArchivableTargetBundleIDToEntitlements() (map[string]serialized.Object, error) {
	targets := append([]xcodeproj.Target{p.MainTarget}, p.MainTarget.DependentExecutableProductTargets(false)...)

	entitlementsByBundleID := map[string]serialized.Object{}

	for _, target := range targets {
		bundleID, err := p.TargetBundleID(target.Name, p.Configuration)
		if err != nil {
			return nil, fmt.Errorf("Failed to get target (%s) bundle id: %s", target.Name, err)
		}

		entitlements, err := p.targetEntitlements(target.Name, p.Configuration)
		if err != nil && !serialized.IsKeyNotFoundError(err) {
			return nil, fmt.Errorf("Failed to get target (%s) bundle id: %s", target.Name, err)
		}

		entitlementsByBundleID[bundleID] = entitlements
	}

	return entitlementsByBundleID, nil
}

// UsesXcodeAutoCodeSigning checks the project uses automatically managed code signing
// It checks the main target "ProvisioningStyle" attribute first then the "CODE_SIGN_STYLE" for the provided configuration
// It returns true if the project uses automatically code signing
func UsesXcodeAutoCodeSigning(xcProj xcodeproj.XcodeProj, mainTarget xcodeproj.Target, config string) (bool, error) {
	settings, err := xcProj.TargetBuildSettings(mainTarget.Name, config)
	if err != nil {
		return false, fmt.Errorf("failed to fetch project settings (%s), error: %s", xcProj.Path, err)
	}

	if provStle, err := settings.String("ProvisioningStyle"); err != nil && !serialized.IsKeyNotFoundError(err) {
		return false, err
	} else if provStle == "Automatic" {
		return true, nil
	}

	for _, buildConf := range mainTarget.BuildConfigurationList.BuildConfigurations {
		if buildConf.Name != config {
			continue
		}

		if signStyle, err := buildConf.BuildSettings.String("CODE_SIGN_STYLE"); err != nil && !serialized.IsKeyNotFoundError(err) {
			return false, err
		} else if signStyle == "Automatic" {
			return true, nil
		}
	}
	return false, nil
}

// Get the platform (PLATFORM_DISPLAY_NAME) - iOS, tvOS, macOS
func platform(xcproj xcodeproj.XcodeProj, mainTarget xcodeproj.Target, configurationName string) (Platform, error) {
	settings, err := xcproj.TargetBuildSettings(mainTarget.Name, configurationName)
	if err != nil {
		return "", fmt.Errorf("failed to fetch project settings (%s), error: %s", xcproj.Path, err)
	}

	platformDisplayName, err := settings.String("PLATFORM_DISPLAY_NAME")
	if err != nil {
		return "", fmt.Errorf("no PLATFORM_DISPLAY_NAME config found for (%s) target", mainTarget.Name)
	}

	if platformDisplayName != string(IOS) && platformDisplayName != string(MacOS) && platformDisplayName != string(TVOS) {
		return "", fmt.Errorf("not supported platform. Platform (PLATFORM_DISPLAY_NAME) = %s", platformDisplayName)
	}
	return Platform(platformDisplayName), nil
}

func configuration(configurationName string, scheme xcscheme.Scheme, xcproj xcodeproj.XcodeProj) (string, error) {
	defaultConfiguration := scheme.ArchiveAction.BuildConfiguration
	var configuration string
	if configurationName == "" || configurationName == defaultConfiguration {
		configuration = defaultConfiguration
	} else if configurationName != defaultConfiguration {
		for _, target := range xcproj.Proj.Targets {
			var configNames []string
			for _, conf := range target.BuildConfigurationList.BuildConfigurations {
				configNames = append(configNames, conf.Name)
			}
			if !sliceutil.IsStringInSlice(configurationName, configNames) {
				return "", fmt.Errorf("build configuration (%s) not defined for target: (%s)", configurationName, target.Name)
			}
		}
		log.Warnf("Using defined build configuration: %s instead of the scheme's default one: %s", configurationName, defaultConfiguration)
		configuration = configurationName
	}

	return configuration, nil
}

// mainTargetOfScheme return the main target
func mainTargetOfScheme(proj xcodeproj.XcodeProj, scheme string) (xcodeproj.Target, error) {
	projTargets := proj.Proj.Targets
	sch, ok := proj.Scheme(scheme)
	if !ok {
		return xcodeproj.Target{}, fmt.Errorf("Failed to found scheme (%s) in project", scheme)
	}

	var blueIdent string
	for _, entry := range sch.BuildAction.BuildActionEntries {
		if entry.BuildableReference.IsAppReference() {
			blueIdent = entry.BuildableReference.BlueprintIdentifier
			break
		}
	}

	// Search for the main target
	for _, t := range projTargets {
		if t.ID == blueIdent {
			return t, nil

		}
	}
	return xcodeproj.Target{}, fmt.Errorf("failed to find the project's main target for scheme (%s)", scheme)
}

// findBuiltProject returns the Xcode project which will be built for the provided scheme
func findBuiltProject(pth, schemeName, configurationName string) (xcodeproj.XcodeProj, string, error) {
	var scheme xcscheme.Scheme
	var schemeContainerDir string

	if xcodeproj.IsXcodeProj(pth) {
		project, err := xcodeproj.Open(pth)
		if err != nil {
			return xcodeproj.XcodeProj{}, "", err
		}

		var ok bool
		scheme, ok = project.Scheme(schemeName)
		if !ok {
			return xcodeproj.XcodeProj{}, "", fmt.Errorf("no scheme found with name: %s in project: %s", schemeName, pth)
		}
		schemeContainerDir = filepath.Dir(pth)
	} else if xcworkspace.IsWorkspace(pth) {
		workspace, err := xcworkspace.Open(pth)
		if err != nil {
			return xcodeproj.XcodeProj{}, "", err
		}

		var containerProject string
		scheme, containerProject, err = workspace.Scheme(schemeName)
		if err != nil {
			return xcodeproj.XcodeProj{}, "", fmt.Errorf("no scheme found with name: %s in workspace: %s, error: %s", schemeName, pth, err)
		}
		schemeContainerDir = filepath.Dir(containerProject)
	} else {
		return xcodeproj.XcodeProj{}, "", fmt.Errorf("unknown project extension: %s", filepath.Ext(pth))
	}

	if configurationName == "" && scheme.ArchiveAction.BuildConfiguration == "" {
		return xcodeproj.XcodeProj{}, "", fmt.Errorf("no configuration provided nor default defined for the scheme's (%s) archive action", schemeName)
	} else if configurationName == "" {
		configurationName = scheme.ArchiveAction.BuildConfiguration
	}

	var archiveEntry xcscheme.BuildActionEntry
	for _, entry := range scheme.BuildAction.BuildActionEntries {
		if entry.BuildForArchiving != "YES" || !entry.BuildableReference.IsAppReference() {
			continue
		}
		archiveEntry = entry
		break
	}

	if archiveEntry.BuildableReference.BlueprintIdentifier == "" {
		return xcodeproj.XcodeProj{}, "", fmt.Errorf("archivable entry not found")
	}

	projectPth, err := archiveEntry.BuildableReference.ReferencedContainerAbsPath(schemeContainerDir)
	if err != nil {
		return xcodeproj.XcodeProj{}, "", err
	}

	project, err := xcodeproj.Open(projectPth)
	if err != nil {
		return xcodeproj.XcodeProj{}, "", err
	}

	return project, scheme.Name, nil
}

// ProjectTeamID returns the development team's ID
// If there is mutlitple development team in the project (different team for targets) it will return an error
// It returns the development team's ID
func (p ProjectHelper) ProjectTeamID(config string) (string, error) {
	var teamID string

	for _, target := range p.Targets {
		currentTeamID, err := targetTeamID(p.XcProj, target.Name, config)
		if err != nil {
			// Do nothing
		}
		log.Debugf("%s target build settings team id: %s", target.Name, currentTeamID)

		if currentTeamID == "" {
			log.Warnf("no DEVELOPMENT_TEAM build settings found for target: %s, checking target attributes...", target.Name)

			targetAttributes, err := p.XcProj.Proj.Attributes.TargetAttributes.Object(target.ID)
			if err != nil {
				return "", fmt.Errorf("failed to parse target %s target attributes, error: %s", target.ID, err)
			}

			targetAttributesTeamID, err := targetAttributes.String("DevelopmentTeam")
			if err != nil && !serialized.IsKeyNotFoundError(err) {
				return "", fmt.Errorf("failed to parse development team for target %s, error: %s", target.ID, err)
			}
			if targetAttributesTeamID == "" {
				log.Warnf("no DevelopmentTeam target attribute found for target: %s", target.Name)
				continue
			}

			currentTeamID = targetAttributesTeamID
		}

		if teamID == "" {
			teamID = currentTeamID
			continue
		}

		if teamID != currentTeamID {
			log.Warnf("target team id: %s does not match to the already registered team id: %s", currentTeamID, teamID)
			teamID = ""
			break
		}
	}

	return teamID, nil

}

// ProjectCodeSignIdentity returns the codesign identity of the project
// If there is mutlitple codesign identity in the project (different identity for targets) it will return an error
// It returns the codesign identity
func (p ProjectHelper) ProjectCodeSignIdentity(config string) (string, error) {
	var codesignIdentity string

	for _, t := range p.Targets {
		targetIdentity, err := targetCodesignIdentity(p.XcProj, t.Name, config)
		if err != nil {
			return "", err
		}

		log.Debugf("%s codesign identity: %s", t.Name, targetIdentity)

		if targetIdentity == "" {
			log.Warnf("no CODE_SIGN_IDENTITY build settings found for target: %s", t.Name)
			continue
		}

		if codesignIdentity == "" {
			codesignIdentity = targetIdentity
			continue
		}

		if !codesignIdentitesMatch(codesignIdentity, targetIdentity) {
			log.Warnf("target codesign identity: %s does not match to the already registered codesign identity: %s", targetIdentity, codesignIdentity)
			codesignIdentity = ""
			break
		}
	}
	return codesignIdentity, nil
}

// 'iPhone Developer' should match to 'iPhone Developer: Bitrise Bot (ABCD)'
func codesignIdentitesMatch(identity1, identity2 string) bool {
	if strings.Contains(strings.ToLower(identity1), strings.ToLower(identity2)) {
		return true
	}
	if strings.Contains(strings.ToLower(identity2), strings.ToLower(identity1)) {
		return true
	}
	return false
}

func targetCodesignIdentity(xcProj xcodeproj.XcodeProj, targatName, config string) (string, error) {
	settings, err := xcProj.TargetBuildSettings(targatName, config)
	if err != nil {
		return "", fmt.Errorf("failed to fetch target (%s) settings, error: %s", targatName, err)
	}
	return settings.String("CODE_SIGN_IDENTITY")
}

func targetTeamID(xcProj xcodeproj.XcodeProj, targatName, config string) (string, error) {
	settings, err := xcProj.TargetBuildSettings(targatName, config)
	if err != nil {
		return "", fmt.Errorf("failed to fetch target (%s) settings, error: %s", targatName, err)
	}

	devTeam, err := settings.String("DEVELOPMENT_TEAM")
	if serialized.IsKeyNotFoundError(err) {
		return devTeam, nil
	}
	return devTeam, err

}

// TargetBundleID returns the target bundle ID
// First it tries to fetch the bundle ID from the `PRODUCT_BUNDLE_IDENTIFIER` build settings
// If it's no available it will fetch the target's Info.plist and search for the `CFBundleIdentifier` key.
// The CFBundleIdentifier's value is not resolved in the Info.plist, so it will try to resolve it by the resolveBundleID()
// It returns  the target bundle ID
func (p ProjectHelper) TargetBundleID(name, conf string) (string, error) {
	settings, err := p.XcProj.TargetBuildSettings(name, conf)
	if err != nil {
		return "", fmt.Errorf("failed to fetch target (%s) settings, error: %s", name, err)
	}

	bundleID, err := settings.String("PRODUCT_BUNDLE_IDENTIFIER")
	if bundleID != "" {
		return bundleID, nil
	}

	log.Debugf("PRODUCT_BUNDLE_IDENTIFIER env not found in 'xcodebuild -showBuildSettings -project %s -target %s -configuration %s command's output", p.XcProj.Path, name, conf)
	log.Debugf("checking the Info.plist file's CFBundleIdentifier property...")

	infoPlistPath, err := settings.String("INFOPLIST_FILE")
	if err != nil {
		return "", fmt.Errorf("failed to find info.plst file, error: %s", err)
	}

	if infoPlistPath == "" {
		return "", fmt.Errorf("failed to to determine bundle id: xcodebuild -showBuildSettings does not contains PRODUCT_BUNDLE_IDENTIFIER nor INFOPLIST_FILE' unless info_plist_path")
	}

	b, err := fileutil.ReadBytesFromFile(infoPlistPath)
	if err != nil {
		return "", fmt.Errorf("failed to read Info.plist, error: %s", err)
	}

	var options map[string]interface{}
	if _, err := plist.Unmarshal(b, &options); err != nil {
		return "", fmt.Errorf("failed to unmarshal Info.plist, error: %s ", err)
	}

	bundleID, ok := options["CFBundleIdentifier"].(string)
	if !ok || bundleID == "" {
		return "", fmt.Errorf("failed to parse CFBundleIdentifier from the Info.plist")
	}

	if !strings.Contains(bundleID, "$") {
		return bundleID, nil
	}

	log.Warnf("CFBundleIdentifier defined with variable: %s, trying to resolve it...", bundleID)
	resolved, err := resolveBundleID(bundleID, settings)
	if err != nil {
		return "", fmt.Errorf("failed to resolve bundle ID, error: %s", err)
	}
	log.Warnf("resolved CFBundleIdentifier: %s", resolved)

	return resolved, nil
}

func resolveBundleID(bundleID string, buildSettings serialized.Object) (string, error) {
	r, err := regexp.Compile(".+[.][$][(].+[:].+[)]*")
	if err != nil {
		return "", err
	}

	if !r.MatchString(bundleID) {
		return "", fmt.Errorf("failed to match regex .+[.][$][(].+[:].+[)]* to %s bundleID", bundleID)
	}

	captures := r.FindString(bundleID)

	prefix := strings.Split(captures, "$")[0]
	envKey := strings.Split(strings.SplitAfter(captures, "(")[1], ":")[0]
	suffix := strings.Join(strings.SplitAfter(captures, ")")[1:], "")

	envValue, err := buildSettings.String(envKey)
	if err != nil {
		return "", fmt.Errorf("failed to find enviroment variable value for key %s, error: %s", envKey, err)
	}
	return prefix + envValue + suffix, nil

}

func (p ProjectHelper) targetEntitlements(name, config string) (serialized.Object, error) {
	o, err := p.XcProj.TargetCodeSignEntitlements(name, config)
	if err != nil && !serialized.IsKeyNotFoundError(err) {
		return nil, err
	}
	return o, nil
}

// TODO
//   def force_code_sign_properties
